hnn_core/network.py

```diff
@@ -984,24 +998,86 @@ def add_connection(self, src_gid, target_gid, loc, receptor,
+                       'int list, range or str')
+        valid_cells = [
+            'L2_pyramidal', 'L2_basket', 'L5_pyramidal', 'L5_basket']
+        # Convert src_gids to list
+        if isinstance(src_gids, int):
+            src_gids = [src_gids]
+        elif isinstance(src_gids, str):
+            _check_option('src_gids', src_gids, valid_cells)
+            src_gids = self.gid_ranges[_long_name(src_gids)]
+
+        # Convert target_gids to list of list, one element for each src_gid
+        if isinstance(target_gids, int):
+            target_gids = [[target_gids] for _ in range(len(src_gids))]
+        elif isinstance(target_gids, str):
+            _check_option('target_gids', target_gids, valid_cells)
+            target_gids = [list(self.gid_ranges[_long_name(target_gids)])
+                           for _ in range(len(src_gids))]
+        elif isinstance(target_gids, range):
+            target_gids = [list(target_gids) for _ in range(len(src_gids))]
+        elif isinstance(target_gids, list) and all(isinstance(t_gid, int)
+                                                   for t_gid in target_gids):
+            target_gids = [target_gids for _ in range(len(src_gids))]
+
+        # Validate each target list - src pairs.
+        # set() used to avoid redundant checks.
+        target_set = set()
+        for target_src_pair in target_gids:
+            _validate_type(target_src_pair, list, 'target_gids[idx]',
+                           'list or range')
+            for target_gid in target_src_pair:
+                target_set.add(target_gid)
+        target_type = self.gid_to_type(target_gids[0][0])
+        for target_gid in target_set:
+            _validate_type(target_gid, int, 'target_gid', 'int')
+            # Ensure gids in range of Network.gid_ranges
+            gid_type = self.gid_to_type(target_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'target_gid {target_gid}''not in net.gid_ranges')
+            elif gid_type != target_type:
+                raise AssertionError(
+                    'All target_gids must be of the same type')
+        conn['target_type'] = target_type
+        conn['num_targets'] = len(target_set)
+
+        if len(target_gids) != len(src_gids):
+            raise AssertionError('target_gids must have a list for each src.')
+
+        # Format gid_pairs and add to conn dictionary
+        gid_pairs = dict()
+        src_type = self.gid_to_type(src_gids[0])
+        for src_gid, target_src_pair in zip(src_gids, target_gids):
+            _validate_type(src_gid, int, 'src_gid', 'int')
+            gid_type = self.gid_to_type(src_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'src_gid {src_gid} not in net.gid_ranges')
+            elif gid_type != src_type:
+                raise AssertionError('All src_gids must be of the same type')
```

Comment:
a lot of these checks are not tested below. But let's move on. Future PR!

Reply:
These error messages are not explicitly matched, but these are checked here:
https://github.com/ntolley/hnn-core/blob/d81205eaf5fd4b7c285337b7b2afd9411c35b920/hnn_core/tests/test_network.py#L203

Agreed though, a future PR should make this clearer in the test code.